File: python/three_sum.py
class Solution(object):
    # O(n^2) time | O(n) space
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()

        for i, a in enumerate(nums):
            if a > 0:
                break
            if i > 0 and a == nums[i - 1]:
                continue
                
            l = i + 1
            r = len(nums) - 1

            while l < r:
                b = nums[l]
                c = nums[r]
                sum = a + b + c
                if sum > 0:
                    r = r - 1
                elif sum < 0:
                    l = l + 1
                else:
                    res.append([a, b, c])
                    l += 1
                    r -= 1
                    while nums[l] == nums[l - 1] and l < r:
                        l = l + 1
        
        return res
    
solution = Solution()

nums = [-1,0,1,2,-1,-4]
print(solution.threeSum(nums))

File: src/main/java/november_2025/ThreeSum.java
package november_2025;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {

    public static void main(String[] args) {
        int[] nums = {-1,0,1,2,-1,-4};

        ThreeSum threeSum = new ThreeSum();
        List<List<Integer>> result = threeSum.threeSum(nums);
        System.out.println(result);

    }

    // O(n^2) time | O(n) space
    public List<List<Integer>> threeSum(int[] nums) {
        if (nums.length < 3) {
            return new ArrayList<>();
        }
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            if (nums[i] > 0) {
                break;
            }
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int num1 = nums[i];
            int l = i + 1;
            int r = nums.length - 1;

            while (l < r) {
                int num2 = nums[l];
                int num3 = nums[r];
                int sum = num1 + num2 + num3;
                if (sum == 0) {
                    result.add(Arrays.asList(num1, num2, num3));
                    l++;
                    r--;
                    while (l < r && nums[l] == nums[l - 1]) {
                        l++;
                    }
                } else if (sum > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }

        return result;
    }


}
